### ISSUE TITLE
Inconsistent behaviour of `skip` option in `useQuery` leading to stale result

### ISSUE THREAD

**Issue by hangindev** [apollographql/apollo-client#11631]
### Issue Description

Consider the following component:

```jsx
function Component({ id, skip }) {
  const { data } = useQuery(GET_DATA, { variables: { id }, skip });
  return data;
}
```

If the variables change while the query is being skipped, I would expect that, when the component re-renders with `skip = false`, the query is made with up-to-date variables and `data` would then reflect the GraphQL query result with those variables applied. However, in some cases (see link to reproduction), the data remains stale and the query is never fetched.

### Link to Reproduction

https://codesandbox.io/p/sandbox/dazzling-violet-9p6hlz?file=%2Fsrc%2FApp.tsx%3A66%2C25&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt5wnube00063b6mxv8wopq8%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt5wnube00023b6mi43bzbp0%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt5wnube00033b6md4ruwwd1%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt5wnube00053b6mk3eem8bs%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt5wnube00023b6mi43bzbp0%2522%253A%257B%2522id%2522%253A%2522clt5wnube00023b6mi43bzbp0%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt5xzwjd00023b6ml1sfth05%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A66%252C%2522startColumn%2522%253A25%252C%2522endLineNumber%2522%253A66%252C%2522endColumn%2522%253A25%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt5xzwjd00023b6ml1sfth05%2522%257D%252C%2522clt5wnube00053b6mk3eem8bs%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt5wnube00043b6m665g1y6g%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clt5wnube00053b6mk3eem8bs%2522%252C%2522activeTabId%2522%253A%2522clt5wnube00043b6m665g1y6g%2522%257D%252C%2522clt5wnube00033b6md4ruwwd1%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clt5wnube00033b6md4ruwwd1%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Afalse%252C%2522sidebarPanelSize%2522%253A15%257D

### Reproduction Steps

1. Click "Toggle View" twice
2. Change character
3. Click "Toggle View" again
4. Observe stale view data

### `@apollo/client` version

3.9.5

**Comment by phryneas** [comment]
Thank you for the reproduction!

Just some notes:

From digging a bit, it seems that in the `fetchPolicy: "standby"` case, `updateWatch` is not called, which results in `updateLastDiff` being called without `options` for the wrong variables - and without `options`, we cannot identify that the variables are actually wrong. 
As a result, the outdated result gets merged in with the new variables.

**Comment by adrienharnay** [comment]
Hello, I have just come across this issue as well, but without using `fetchPolicy: "standby"`.

I have noticed that when I execute a query, then set `skip: true`, then call `client.clearStore()` or `client.cache.updateQuery()`, then set `skip: false`, `data` will be returned as the value it was before setting `skip: true`. I would expect `data` to match what is actually in the Apollo cache.

Here is a reproduction project: https://github.com/Brigad/apollo-cache-bug-repro with detailed reproduction steps to highlight the behavior I am describing.

Please let met know if you would prefer I open a new issue, this one having been stale for a while.

**Comment by phryneas** [comment]
@adrienharnay I believe this specific behaviour should be solved in Apollo Client 4.0 - we're targeting a RC release this week, but please already try the alpha and report back.

I've tested it in your reproduction and there it works as you expect: https://codesandbox.io/p/devbox/sleepy-sanderson-73rxmc

Please note that `Provider`, `useQuery` etc. are only available in `@apollo/client/react` from 4.0 on, so you'll have to at least change some imports.

**Comment by adrienharnay** [comment]
Indeed, I have tried the alpha on the codesanbox you have shared and it works as I would expect! Thanks for the quick reply.

**Comment by jerelmiller** [comment]
Hey @adrienharnay 👋 

We've [officially released Apollo Client 4.0](https://www.apollographql.com/blog/announcing-apollo-client-4-0) and as a result, this issue should be fixed. As @phryneas demonstrated in an early alpha build, this issue was addressed.

Please see the [migration guide](https://www.apollographql.com/docs/react/migrating/apollo-client-4-migration) for instructions on upgrading to 4.0. Since this has been fixed and 4.0 is officially released, I'm going to go ahead and close this as completed. If you see issues after upgrading, let us know and we'd be happy to take a look. Thanks!
